File: src/com/gdb/tests/TestAbstractAccount.java
package com.gdb.tests;
import com.gdb.domain.*;
import com.gdb.exceptions.*;

public class TestAbstractAccount {
    public static void main(String[] args) {
        System.out.println("=== Activity 9: Abstract Account & Template Pattern ===");
        SavingsAccount sa = new SavingsAccount("SA1001", "Rahul", 30, 10000.0, "ACTIVE", "1111", 1000.0, 4.0);
        CurrentAccount ca = new CurrentAccount("CA1001", "Business", 40, 5000.0, "ACTIVE", "2222", 25000.0);
        FixedDepositAccount fda = new FixedDepositAccount("FD1001", "Ravi", 50, 50000.0, "ACTIVE", "3333", 12, 6.5);
        
        System.out.print("[Savings] Withdraw 2000: ");
        try {
            sa.withdraw(2000.0, "1111");
            System.out.println("SUCCESS | Balance: Rs " + sa.getBalance());
        } catch (Exception e) {
            System.out.println("FAILED");
        }

        System.out.print("[Savings] Withdraw below min balance: ");
        try {
            sa.withdraw(8000.0, "1111");
            System.out.println("FAILED");
        } catch (MinimumBalanceViolationException e) {
            System.out.println("Caught MinimumBalanceViolationException [PASS]");
        } catch (Exception e) {
            System.out.println("FAILED");
        }

        System.out.print("[Current] Overdraft debit: ");
        try {
            ca.withdraw(8000.0, "2222");
            System.out.println("SUCCESS | Balance: Rs " + ca.getBalance());
        } catch (Exception e) {
            System.out.println("FAILED");
        }

        System.out.print("[FixedDeposit] Premature debit: ");
        try {
            fda.withdraw(5000.0, "3333");
            System.out.println("FAILED");
        } catch (AccountException e) {
            System.out.println("Caught AccountException [PASS]");
        }

        System.out.println("Template method pattern executed successfully!");
    }
}

File: gdb/tests/test_abstract_account.py

from gdb.domain.savings_account import SavingsAccount
from gdb.domain.current_account import CurrentAccount
from gdb.domain.fixed_deposit_account import FixedDepositAccount
from gdb.exceptions import *

print("=== Activity 10: Abstract Account Test ===")
sa = SavingsAccount("SA1", "A", 20, 5000.0, "ACTIVE", "1234", 1000.0, 4.0)
ca = CurrentAccount("CA1", "B", 30, 5000.0, "ACTIVE", "1234", 10000.0)
fda = FixedDepositAccount("FD1", "C", 40, 50000.0, "ACTIVE", "1234", 12, 6.5)

try:
    sa.withdraw(4500.0, "1234")
    print("FAILED")
except MinimumBalanceViolationException:
    print("Caught MinimumBalanceViolationException [PASS]")
except Exception as e:
    print(f"FAILED {e}")

print("Test complete!")
